### PR TITLE
Add support for CORS

### PR DESCRIPTION
**Pull request by imishravmw**
**Issue Ref**: [Issue number related to this PR or None]
 934
**Description**: 
Support for cors flag for http trigger

[PR Description]
cors-enable flag is a boolean flag, by default it is false to be backward compatible.
if cors-enable flag is true, then one can specify cors-domain  which will set that only from these cors-domain, cors will be supported.
when cors-enable flag is true while creating HTTP trigger then it adds required annotations in ingress resource to enable cors

**TODOs**:
 - [x] Ready to review
 - [x] Automated Tests
 - [ ] Docs

### CODE REVIEW

File: pkg/utils/k8sutil.go [hunk 1]
```diff
@@ -19,19 +19,18 @@ package utils
...
 	kubelessApi "github.com/kubeless/http-trigger/pkg/apis/kubeless/v1beta1"
+	"github.com/sirupsen/logrus"
 	"k8s.io/api/extensions/v1beta1"
 	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/clientcmd"
-
-	"github.com/sirupsen/logrus"
+	"net/url"
```
Comment (andresmgot) [inline pkg/utils/k8sutil.go]: the core imports should be in a block separated by a empty line from other libraries

File: script/libtest.bash [hunk 2]
```diff
@@ -457,6 +457,23 @@ create_http_trigger(){
     fi
     eval $command
 }
+
+create_http_trigger_cors(){
```
Comment (andresmgot) [inline script/libtest.bash]: you don't need to define a different function, you can add a parameter to `create_http_trigger` to add the cors parameter

File: tests/integration-tests-http.bats [hunk 3]
```diff
@@ -64,6 +64,19 @@ load ../script/libtest
...
+    deploy_function get-python
+    verify_function get-python
+	create_http_trigger_cors get-python  "cors_enable" "test.domain"
+	verify_http_trigger_cors get-python $(minikube ip) "Access-Control-Allow-Origin: *" "test.domain" "test.domain"
+	delete_http_trigger get-python
+	verify_clean_object httptrigger ing-get-python
+	verify_clean_object ingress ing-get-python
+}
+
+
```
Comment (andresmgot) [inline tests/integration-tests-http.bats]: you can leave this with a single empty line

**Review by andresmgot — COMMENTED** [review]
Hi @imishravmw, thanks for this PR!

I have a suggestion though to make this feature more extensible (and easy to implement). In my opinion we can just forward the annotations from the http trigger object to the ingress so we cover this and any other future request. It also allow people not using the `nginx` ingress controller to define the annotations that they want:

```patch
	ingressAnnotations := make(map[string]string)
+	if len(httpTriggerObj.ObjectMeta.Annotations) > 0 {
+	  ingressAnnotations = httpTriggerObj.ObjectMeta.Annotations
+       }
```

Another thing, to fix the CI a change is required in the file `.circleci/config.yml`. You  need to add these two lines:

```patch
jobs:
  build:
    <<: *defaults
    docker:
      - image: circleci/golang:1.9 
    steps:
      - checkout
      - <<: *exports
+      # go build creates files under /usr/local/go where the running user doesn't have permissions to write
+      - run: sudo chown -R circleci:circleci /usr/local/go/
      - run: make bootstrap
      - run: mkdir build-manifests
```

**Comment by imishravmw** [comment]
> ```diff
> httpTriggerObj.ObjectMeta.Annotations
> ```

But using this approach we expect user to know what annotation to be applied on HTTPTrigger object   , if one creates HttpTrigger object using yaml definition then one can do this ( rightly this will be more extensible). 

 but while creating trigger with kubeless CLI as well, then don't we want to give user a flag to enable cors.  or we give a way to pass metadata as well. Then it will be fine.

in all ingress controllers, there is a way to enabled cors using annotation, 
is there a way to know  which ingress controller is being used

**Comment by imishravmw** [comment]
Please see my latest changes, 
It copies annotation from metadata to http trigger object , and if using kubeless cli someone specifies cors-enable flag then also it applies cors enable annotation .

Still circleci build is failing because of below error
Build Kubeless Components successful. Program saved at /home/circleci/.go_workspace/bin
kubecfg show -U https://raw.githubusercontent.com/kubeless/runtimes/master -o yaml kubeless.jsonnet > kubeless.yaml.tmp
ERROR unknown shorthand flag: 'U' in -U
Makefile:36: recipe for target 'kubeless.yaml' failed

Can you let me know what could be problem, @andresmgot

File: script/libtest.bash [hunk 4]
```diff
@@ -438,7 +438,7 @@ create_http_trigger(){
     local func=${1:?}; shift
     local domain=${1-""};
     local subpath=${2-""};
-    local basicauth=${3-""};
+    local var3=${3-""};
```
Comment (andresmgot) [inline script/libtest.bash]: can use two different arguments for both? if you want to leave an empty parameter you can use `""` like:

```
create_http_trigger foo foo.com /foo "" cors
```

File: script/libtest.bash [hunk 5]
```diff
@@ -449,14 +449,18 @@ create_http_trigger(){
     if [ -n "$subpath" ]; then
         command="$command --path ${subpath}"
     fi
-    if [ -n "$basicauth" ]; then
+    if [[ $var3 = "basic-auth" ]]; then
         command="$command --basic-auth-secret ${basicauth}"
```
Comment (andresmgot) [inline script/libtest.bash]: this is now broken because `${basicauth}` is not defined.

**Review by andresmgot — COMMENTED** [review]
> but while creating trigger with kubeless CLI as well, then don't we want to give user a flag to enable cors. or we give a way to pass metadata as well. Then it will be fine.

That's a fair point. We can have both.

> Can you let me know what could be problem, @andresmgot

Sorry, a build tool version was outdated. I have updated it in your branch. Remember to pull those changes before submitting new commits.

There are two errors now in the tests. The first one is simple, you just need to restore the function for using `basicauth`.

The second is a bit more complex. We have now a circular dependency. This build uses the `master` branch of Kubeless. In that branch the flag `--enable-cors` is still not present so the test is failing here. The problem is that in Kubeless the PR https://github.com/kubeless/kubeless/pull/1008 is also failing because it needs the changes here.

My suggestion is to avoid here using the flag `--enable-cors` as a workaround. So you can change in the test the line `create_http_trigger get-python "test.domain" "get-python" "cors"` for using directly the YAML content. Something like:

```
kubectl apply -f - -o yaml << EOF
apiVersion: kubeless.io/v1beta1
kind: HTTPTrigger
...
EOF
```

File: script/libtest.bash (hunk 4 above)
Comment (imishravmw) [reply script/libtest.bash]: will change it like this.

**Comment by imishravmw** [comment]
@andresmgot please see now ,I have removed circular dependency in tests,  build is passing now.
Will make change in kubeless PR as well, 
Will add in kubeless doc regarding how to enable CORS
Thanks for help.

File: examples/python/cors-trigger.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: kubeless.io/v1beta1
+kind: HTTPTrigger
+metadata:
+  name: ing-cors-trigger
+spec:
+  function-name: get-python
+  host-name: test.domain
+  path: echo
+  gateway: nginx
+  cors-enable: true
```
Comment (andresmgot) [inline examples/python/cors-trigger.yaml]: can you add a new empty line after this?

File: script/libtest.bash [hunk 7]
```diff
@@ -455,8 +456,13 @@ create_http_trigger(){
     if [ -n "$gateway" ]; then
         command="$command --gateway ${gateway}"
     fi
-    eval $command
+    if [ -n "$cors" ]; then
+        make -sC examples create-cors-trigger
```
Comment (andresmgot) [inline script/libtest.bash:461]: can you add a command here to explain why this behavior is different? Something like:
```
# TODO: use --cors-enable flag when there is a new version of Kubeless CLI
```

**Review by andresmgot — APPROVED** [review]
Thanks for the changes @imishravmw! I just have a couple of minor comments now

**Comment by imishravmw** [comment]
@andresmgot  changes are done and build has passed.

**Comment by imishravmw** [comment]
@andresmgot can you merge this PR , as my second PR is dependent on this merge

**Comment by andresmgot** [comment]
Thanks! I am merging this.

**Pull request merged**